src/Controller/EditOrderAction.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Setono\SyliusOrderEditPlugin\Controller;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Setono\SyliusOrderEditPlugin\Checker\PostUpdateChangesCheckerInterface;
+use Setono\SyliusOrderEditPlugin\Exception\NewOrderWrongTotalException;
+use Setono\SyliusOrderEditPlugin\Preparer\OrderPreparerInterface;
+use Setono\SyliusOrderEditPlugin\Processor\UpdatedOrderProcessorInterface;
+use Setono\SyliusOrderEditPlugin\Provider\UpdatedOrderProviderInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+final class EditOrderAction
+{
+    public function __construct(
+        private readonly OrderPreparerInterface $oldOrderProvider,
+        private readonly UpdatedOrderProviderInterface $updatedOrderProvider,
+        private readonly UpdatedOrderProcessorInterface $updatedOrderProcessor,
+        private readonly PostUpdateChangesCheckerInterface $postUpdateChangesChecker,
+        private readonly UrlGeneratorInterface $router,
+        private readonly RequestStack $requestStack,
+        private readonly EntityManagerInterface $entityManager,
+    ) {
+    }
+
+    public function __invoke(Request $request, int $id): Response
+    {
+        $order = $this->oldOrderProvider->prepareToUpdate($id);
+
+        $oldOrder = clone $order;
+        $updatedOrder = $this->updatedOrderProvider->provideFromOldOrderAndRequest($order, $request);
+
+        try {
+            $this->updatedOrderProcessor->process($updatedOrder);
+            $this->postUpdateChangesChecker->check($oldOrder, $updatedOrder);
```

Comment:
What's the reason for not doing this as a form validator?

Reply:
As it's a business logic check, I feel it's just easier from the DX perspective to use a custom service with a custom exception. Also validator enforces on us some contract that seems irrelevant for this vase (validation groups, separate classes for constraints) and could we even easily pass two values to the validator (old order and updated order)? 
Again, as most of the conversations we have here, it's a matter of taste and conventions, not a real problem from the application point of view 😄 If you want me to change it to the validator - let me know 🖖

Reply:
There's a lot of problems with doing it using the validator, I agree. Let's keep it this way